tensorboard/plugins/debugger_v2/debug_data_multiplexer.py

```diff
@@ -0,0 +1,122 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""A wrapper around DebugDataReader used for retrieving tfdbg v2 data."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+# Dummy run name for the debugger.
+# Currently, the `DebuggerV2ExperimentMultiplexer` class is tied to a single
+# logdir, which holds at most one DebugEvent file set in the tfdbg v2 (tfdbg2
+# for short) format.
+# TODO(cais): When tfdbg2 allows there to be multiple DebugEvent file sets in
+# the same logdir, replace this magic string with actual run names.
+DEFAULT_DEBUGGER_RUN_NAME = "__default_debugger_run__"
+
+
+class DebuggerV2EventMultiplexer(object):
+    """A class used for accessing tfdbg v2 DebugEvent data on local filesystem.
+
+    This class is a short-term hack, mirroring the EventMultiplexer for the main
+    TensorBoard plugins (e.g., scalar, histogram and graphs.) As such, it only
+    implements the methods relevant to the Debugger V2 pluggin.
+
+    TODO(cais): Integrate it with EventMultiplexer and use the integrated class
+    from MultiplexerDataProvider for a single path of accessing debugger and
+    non-debugger data.
+    """
+
+    def __init__(self, logdir):
+        """Constructor for the `DebugEventMultiplexer`.
+
+        Args:
+          logdir: Path to the directory to load the tfdbg v2 data from.
+        """
+        self._logdir = logdir
+        # TODO(cais): Start off a reading thread here.
+
+    def FirstEventTimestamp(self, run):
+        """Return the timestamp of the first DebugEvent of the given run.
+
+        This may perform I/O if no events have been loaded yet for the run.
+
+        Args:
+          run: A string name of the run for which the timestamp is retrieved.
+            This currently must be hardcoded as `DEFAULT_DEBUGGER_RUN_NAME`,
+            as each logdir contains at most one DebugEvent file set (i.e., a
+            run of a tfdbg2-instrumented TensorFlow program.)
+
+        Returns:
+          The wall_time of the first event of the run, which will be in seconds
+          since the epoch as a `float`.
+        """
+        if run != DEFAULT_DEBUGGER_RUN_NAME:
+            raise ValueError(
+                "Expected run name to be %s, but got %s"
+                % (DEFAULT_DEBUGGER_RUN_NAME, run)
+            )
+        reader = None
+        try:
+            from tensorflow.python.debug.lib import debug_events_reader
+
+            # TODO(cais): Switch DebugDataReader when it supports metadata.
+            reader = debug_events_reader.DebugEventsReader(self._logdir)
+            metadata_iterator, _ = reader.metadata_iterator()
+            return next(metadata_iterator).wall_time
+        finally:
+            reader.close()
```

Comment:
Won’t this be a `UnboundLocalError` if the reader fails to initialize
(which, looking at the implementation, seems possible)? Should this code
use the same `with reader` pattern as the other call site?

Reply:
Good catch. Done.